File: src/difference-array/2528-maximize-the-minimum-powered-city/cpp/src/solution.cpp
#include "solution.hpp"

#include <algorithm>
#include <numeric>
#include <vector>

/**
 * Finds the largest minimum power achievable after adding at most k new stations.
 * Binary searches the answer, using a difference array to simulate coverage additions.
 */
long long maxPower(const std::vector<int>& stations, int r, long long k)
{
  int n = stations.size();
  std::vector<long long> cnt(n + 1); // base difference array for window coverage
  for (int i = 0; i < n; i++)
  {
    int left = std::max(0, i - r);
    int right = std::min(n, i + r + 1);
    cnt[left] += stations[i];
    cnt[right] -= stations[i];
  }

  auto check = [&](long long val) -> bool
  {
    std::vector<long long> diff = cnt; // mutable difference array while we add power
    long long sum = 0;
    long long remaining = k;

    for (int i = 0; i < n; i++)
    {
      sum += diff[i];
      if (sum < val)
      {
        long long add = val - sum;
        if (remaining < add)
        {
          return false;
        }
        remaining -= add;
        int end = std::min(n, i + 2 * r + 1); // new station affects cities up to this index
        diff[end] -= add;
        sum += add;
      }
    }
    return true;
  };

  long long lo = std::ranges::min(stations);
  long long hi = accumulate(stations.begin(), stations.end(), 0LL) + k;
  long long res = 0;
  while (lo <= hi)
  {
    long long mid = lo + (hi - lo) / 2;
    if (check(mid))
    {
      res = mid;
      lo = mid + 1;
    }
    else
    {
      hi = mid - 1;
    }
  }
  return res;
}

File: src/difference-array/2528-maximize-the-minimum-powered-city/typescript/solution.ts
/**
 * Binary-searches the best achievable minimum power after adding at most k stations.
 * Builds a difference array to quickly simulate the circular coverage window for each city.
 */
export function maxPower(stations: number[], r: number, k: number): number {
  const n: number = stations.length;
  const cnt: number[] = new Array(n + 1).fill(0); // coverage difference array

  // Build the base difference array representing current window coverage
  for (let i = 0; i < n; i++) {
    const left: number = Math.max(0, i - r);
    const right: number = Math.min(n, i + r + 1);
    cnt[left] += stations[i];
    cnt[right] -= stations[i];
  }

  const check = (val: number): boolean => {
    // diff keeps the mutable prefix sums as we inject new stations
    const diff: number[] = [...cnt];
    let sum: number = 0;
    let remaining: number = k;

    for (let i = 0; i < n; i++) {
      sum += diff[i];
      if (sum < val) {
        const add: number = val - sum;
        if (remaining < add) {
          return false;
        }
        remaining -= add;
        const end: number = Math.min(n, i + 2 * r + 1); // effect of added stations ends here
        diff[end] -= add;
        sum += add;
      }
    }
    return true;
  };

  let lo: number = Math.min(...stations);
  let hi: number = stations.reduce((a, b) => a + b, 0) + k;
  let res: number = 0;

  while (lo <= hi) {
    const mid: number = Math.floor(lo + (hi - lo) / 2);
    if (check(mid)) {
      res = mid;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  return res;
}
